### PR TITLE
Improved Status Checking

### PR DESCRIPTION
**Pull request by cconard96**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

As GLPI and plugins become more complex and reliant on external services/systems it becomes more likely that a configuration or connectivity issue (In GLPI or the external service) could partially or completely break GLPI or the plugins, and it becomes more necessary to monitor and alert on these issues.
The goal of this PR is to make it easier to monitor GLPI. In some monitoring software, it may be inconvenient to do that with the previous implementation because all status checks were done together and returned as a single result which is slow and inefficient if you only want to check one or two things. The goal is to have the option to break up these checks and allow each "service" to be monitored at differing intervals, under separate services in the monitoring software, skipped completely, etc.

- [x] Break up the different status checks into different "services" (DB, Email, Automatic Actions, etc) which can be checked individually instead of all at once to allow administrators to more easily monitor each one more efficiently. Leaving out the service name or using the name "all" will check everything as before.
- [x] Add command to list all available services
- [x] Add warning level in addition to the existing OK and Problem statuses which can be used to describe a service that works, at least partially, but may not be working as best as it should. There is nothing in the core that uses it now, but it is taken into account for the overall system status calculation in case it is used in the future or if a plugin uses it.
- [x] Pass a parameter to the plugin hook so plugins can add more or less information based on if only public information is requested (using public status endpoint instead of the console with -p flag) or not.

- [x] Provide compatible output for previous full status check with a plain-text format. JSON output should already be the same.

New Commands Examples:
```
$ bin/console system:service:list
db
cas
ldap
imap
mail_collectors
crontasks
filesystem
plugins
```
```
$ bin/console system:status -service db
DB_OK
```
```
$ bin/console system:status -s db -f json
{
    "status": "OK",
    "master": {
        "status": "OK"
    },
    "slaves": {
        "status": "NO_DATA",
        "servers": []
    }
}
```

### CODE REVIEW

**Comment by cconard96** [comment]
Can the plain-text output for the status endpoint and command be deprecated in favor of structured and extensible formats only such as the already implemented JSON?

**Comment by cedric-anne** [comment]
> Can the plain-text output for the status endpoint and command be deprecated in favor of structured and extensible formats only such as the already implemented JSON?

I'm not very familiar with the monitoring tools ecosystem, but it seems to me that modern tools are capable of handling JSON format. So I think it's okay to deprecate this plain-text format, since maintaining it requires extra work.

File: inc/console/system/listservicecommand.class.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+   die("Sorry. You can't access this file directly");
+}
+
+use Glpi\Console\AbstractCommand;
+use Glpi\System\Status\StatusChecker;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class ListServiceCommand extends AbstractCommand {
```
Comment (cedric-anne) [inline inc/console/system/listservicecommand.class.php:45]:
```suggestion
class ListServicesCommand extends AbstractCommand {
```

File: inc/console/system/listservicecommand.class.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+   protected function configure() {
+      parent::configure();
+
+      $this->setName('glpi:system:service:list');
+      $this->setAliases(['system:service:list']);
```
Comment (cedric-anne) [inline inc/console/system/listservicecommand.class.php:51]: I think we should keep the `glpi:x:y` pattern.

```suggestion
      $this->setName('glpi:system:list_services');
      $this->setAliases(['system:list_services']);
```

File: inc/console/system/listservicecommand.class.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      $this->setDescription(__('List system services'));
+      $this->addOption('format', 'f', InputOption::VALUE_OPTIONAL,
```
Comment (cedric-anne) [inline inc/console/system/listservicecommand.class.php:53]: I am not sure that proposing the plain format here is usefull, if we deprecate this format in the `glpi:system:status` command.

File: inc/console/system/listservicecommand.class.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+class ListServicesCommand extends AbstractCommand {
+
+   protected function configure() {
+      parent::configure();
+
+      $this->setName('glpi:system:list_services');
+      $this->setAliases(['system:list_services']);
+      $this->setDescription(__('List system services'));
+      $this->addOption('format', 'f', InputOption::VALUE_OPTIONAL,
+         'Output format [json]', 'plain');
```
Comment (cedric-anne) [inline inc/console/system/listservicecommand.class.php:54]:
```suggestion
```

File: inc/console/system/listservicecommand.class.php [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+   }
+
+   protected function execute(InputInterface $input, OutputInterface $output) {
+
+      $format = strtolower($input->getOption('format'));
+
```
Comment (cedric-anne) [inline inc/console/system/listservicecommand.class.php:60]:
```suggestion
```

File: inc/console/system/listservicecommand.class.php [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+      $services = array_keys(StatusChecker::getServices());
+
+      if ($format === 'json') {
+         $output->writeln(json_encode($services, JSON_PRETTY_PRINT));
+      } else {
+         foreach ($services as $service) {
+            $output->writeln($service);
+         }
+      }
```
Comment (cedric-anne) [inline inc/console/system/listservicecommand.class.php:69]:
```suggestion
      $output->writeln(json_encode($services, JSON_PRETTY_PRINT));
```

**Review by cedric-anne — COMMENTED** [review]
Completely removing alternative format handling for the new command would simplify the code.
I guess most of monitoring tools can deal with JSON format.

**Review by cedric-anne — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Pull request merged**